### ISSUE TITLE
Add react-pdf-native

### ISSUE THREAD

**Issue by jbovenschen** [diegomura/react-pdf#34]
Make it possible to render pdf's inside react-native.
Same way as the react-pdf-dom package works.

**Comment by rbansalrahul6** [comment]
Hi, I recently started working in react native and was looking for a PDF rendering library for it. Somehow I landed here. I really want to contribute for this feature. Can you please guide me on how to proceed since I am also a beginner in open source?

**Comment by diegomura** [comment]
Hi @rbansalrahul6 !
Thanks for reaching out!

So this feature is basically creating some bindings to show pdf's using react-native and react-pdf.
This project is divieded into several npm packages. You can find them under `/packages`. `react-pdf` package has all the shared logic for pdf creation, and then we have one package for each platform we want to support. `react-pdf` basically returns a buffer of the generated PDF document, and then each binding takes that buffer and does with it something platform-specific to show it. You can see that very clear in the `react-pdf/node` bindings: https://github.com/diegomura/react-pdf/blob/master/packages/react-pdf-node/src/index.js

So all we have to do about native is something very similar. There are already some libraries out there to render PDF documents in native. We can use one of those for start, and use react-pdf to actually create the buffer. However, when I tried something with invalid dependencies showed up, and I ever worked on this again (not a big react-native user also). 

So what you can do is forking the project, create a new branch there and start figuring out if you can create that new binding. Please ask me any doubt and I'll try to help 😉

**Comment by rbansalrahul6** [comment]
Thanks for guiding me. I'll try to follow it. Can you tell me the currently available libraries for rendering PDF in react native. I couldn't find any good one.

**Comment by diegomura** [comment]
Oh. I actually don't know anyone. But I see there are some that does that. Part of the work is searching for that, or build it from scratch haha.
We are using `pdfkit` for the PDF creation

**Comment by diegomura** [comment]
Here you have one that seems nice: https://github.com/wonday/react-native-pdf
We would have to test it though. As you can see on the bottom of the README, it supports passing by a base64 document as parameter

**Comment by rbansalrahul6** [comment]
I think [https://github.com/wonday/react-native-pdf](url) is for viewing PDFs, not for rendering them

**Comment by diegomura** [comment]
Yes. It is. But we can use `react-pdf` to create the pdf (in base 64 format for example) and then use `react-native-pdf` to show it. right? It's just an idea. I'm open to any other solution

**Comment by bansal-rahul** [comment]
Ok. I'll look into it

**Comment by rbansalrahul6** [comment]
Ok. I'll look into it and let you know if something works

**Comment by jbovenschen** [comment]
@rbansalrahul6 Do you have any progress on this issue? Currently I'm rewriting the pdfkit library to be more web compatible API's. Would make it a lot easier to get this package working.

Will push my progress asap to the https://github.com/react-pdf/pdfkit fork.

**Comment by dancomanlive** [comment]
Any updates on this issue?

**Comment by diegomura** [comment]
No. Sorry

**Comment by luizcieslak** [comment]
I was able to solve this by using [renderToString](https://reactjs.org/docs/react-dom-server.html#rendertostring) method in a `<Component />`  and using that as a input to [react-native-html-to-pdf](https://github.com/christopherdro/react-native-html-to-pdf) library.

**Comment by dancomanlive** [comment]
@luizcieslak Do you have an example?

**Comment by andrioid** [comment]
@luizcieslak How did you use the renderToString method on the output from React PDF?

**Comment by luizcieslak** [comment]
@andrioid @dancomanlive here's the example. Not a fan of it since it mixes `react-dom` and `react-native` dependencies but it works ¯\\_(ツ)_/¯

Let me know if u find another solution!

```js
import { renderToString } from 'react-dom/server' // Component to HTML
import RNHTMLtoPDF from 'react-native-html-to-pdf' // HTML to PDF
import resolveAssetSource from 'react-native/Libraries/Image/resolveAssetSource' // Custom font to HTML

import Example from './Example'

// ...
// in a button click event handler
const htmlstring = renderToString(<Example />)
const options = {
  html: `
        <style> 
          @font-face {
            font-family: opensans;
            src: url(opensans.ttf);
          }
          *{ font-family: opensans;}
        </style>
      <body>
        ${htmlstring}
      </body>`,
  fileName: 'filename',
  directory: 'Documents',
  base64: true,
  fonts: [resolveAssetSource(require('../assets/fonts/opensans.ttf')).uri]
}
try {
  const file = await RNHTMLtoPDF.convert(options)
  // file is a object that has filePath and base64 fields
} catch (error) {
  console.log('catch error convert pdf', error)
}
```

Example component

```js
import React from 'react'
import { StyleSheet } from 'react-native'

const Example = () => (
	<div style={styles.container} />
)

// not sure why but CSS classes didnt work here
const styles = StyleSheet.create({
	container: {
		borderStyle: 'solid',
		borderWidth: 1,
		borderColor: '#CCC',
		borderRadius: 5,
		padding: 10
	}
})
```

**Comment by ravithesun02** [comment]
I am able to generate pdf in react-native using "**nodejs-mobile-react-native**": "^0.8.2"and **react-pdf (renderToStream)** and **rn-fetch-blob-util**.

**Comment by stephanrueschenbaum** [comment]
@ravithesun02 Could you please provide an example repo with your setup? How did you force the metro bundler to load the nodejs supported version of @react-pdf/renderer instead of the browser version?

I try this for days inside an expo project... But I always get this error message:
![image](https://user-images.githubusercontent.com/7265608/202783577-f8f5f6b9-9302-4db5-95b6-6cf9b8dd1bd9.png)

Thanks!!

**Comment by ravithesun02** [comment]
@stephanrueschenbaum  I used nodejs mobile react native package run node on a thread inside react native and then used nodeAPI of react pdf . 
I will try to prepare a example project . 
But I am not sure about expo.

**Comment by stephanrueschenbaum** [comment]
@ravithesun02 Thanks for the reply!! A standard ReactNative project will work as well. But I'm not sure how to build the bridge between NodeJS and ReactNative to finally pass an JSON object from RN to the NodeJS process and then stream the PDF document in the ReactNative view.

I can imagine that you stream the PDF via a channel using the `rn-bridge` package...

Maybe this workaround would be very useful for other users as well, which already invested a lot of time building PDFs using @react-pdf and now want to render these in a mobile application.

Would really appreciate your example!

**Comment by ravithesun02** [comment]
@stephanrueschenbaum , for now here is an example from node side and how i am handling in react-native side 

Node main.js file 
`var rn_bridge = require('rn-bridge');
const React = require('react');
const { renderToStream } = require('@react-pdf/renderer');
const {PrescriptionTemplate} = require('./Templates/PrescriptionTemplate');

// Echo every message received from react-native.
rn_bridge.channel.on('prescriptionDownload', (data) => {
  console.log("[node] Received and event for Prescription Download");
  generatePdf(data);
});
rn_bridge.app.on('pause', (pauseLock) => {
  console.log('[node] app paused.');
  pauseLock.release();
});
rn_bridge.app.on('resume', () => {
  console.log('[node] app resumed.');
});

async function generatePdf(data) {

  try {
    const stream = await renderToStream(<PrescriptionTemplate data={data}/>);
    stream.setEncoding('base64');

    let base64Data='';
    stream.on('data', data=>{
      base64Data+=data;
    })
    stream.on('end',()=>{
      console.log("[node] Sent an event for Prescription Download");
      rn_bridge.channel.post('prescriptionDownload',base64Data);
    })

   
  } catch (error) {
    console.log(error);
    return error;
  }

}

console.log('[NODE] nodejs server started....');`

before calling the main i am preparing a build using babel , here is my package.json from node 
`{
  "name": "sample-node-project",
  "version": "0.0.1",
  "description": "node part of the project",
  "main": "main.js",
  "scripts": {
    "build": "babel src -d lib"
  },
  "dependencies": {
    "@react-pdf/font": "2.2.0",
    "@react-pdf/renderer": "2.0.16",
    "moment": "^2.29.4",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  },
  "resolutions": {
    "@react-pdf/font": "2.2.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.16.0",
    "@babel/core": "^7.16.0",
    "@babel/node": "^7.16.0",
    "@babel/plugin-proposal-optional-chaining": "^7.18.9",
    "@babel/preset-react": "^7.16.0"
  }
}
`


and on componentDidMount 
nodejs.start('lib/main.js');

to send data to pdf template 

`nodejs.channel.addListener(
          'prescriptionDownload',
          (msg) => {
            console.log('.......MESSAGE FROM NODEJS >>>>>>>>>>>>>>>>>>>');
            this.setState({
              prescriptionBase64:msg,
              isPdfDataLoading:false
            })
          },
          this 
        );`

`nodejs.channel.post('prescriptionDownload',temaplateData);`
